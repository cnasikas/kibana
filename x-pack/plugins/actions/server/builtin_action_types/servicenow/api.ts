/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
import {
  PushToServiceApiHandlerArgs,
  HandshakeApiHandlerArgs,
  GetIncidentApiHandlerArgs,
  ExternalServiceApi,
  PushToServiceResponse,
  GetCommonFieldsHandlerArgs,
  GetCommonFieldsResponse,
} from './types';

const handshakeHandler = async ({ externalService, params }: HandshakeApiHandlerArgs) => {};
const getIncidentHandler = async ({ externalService, params }: GetIncidentApiHandlerArgs) => {};

const pushToServiceHandler = async ({
  externalService,
  params,
  secrets,
}: PushToServiceApiHandlerArgs): Promise<PushToServiceResponse> => {
  const { externalId, comments } = params;
  let res: PushToServiceResponse;

  const incident = { ...params, short_description: params.title, comments: params.comment };

  if (externalId != null) {
    res = await externalService.updateIncident({
      incidentId: externalId,
      incident,
    });
  } else {
    res = await externalService.createIncident({
      incident: {
        ...incident,
        caller_id: secrets.username,
      },
    });
  }

  // TODO: should temporary keep comments for a Case usage
  if (comments && Array.isArray(comments) && comments.length > 0) {
    res.comments = [];

    for (const currentComment of comments) {
      await externalService.updateIncident({
        incidentId: res.id,
        incident: {
          ...incident,
          comments: currentComment.comment,
        },
      });
      res.comments = [
        ...(res.comments ?? []),
        {
          commentId: currentComment.commentId,
          pushedDate: res.pushedDate,
        },
      ];
    }
  }
  return res;
};

const getFieldsHandler = async ({
  externalService,
}: GetCommonFieldsHandlerArgs): Promise<GetCommonFieldsResponse> => {
  const res = await externalService.getFields();
  return res;
};

export const api: ExternalServiceApi = {
  getFields: getFieldsHandler,
  getIncident: getIncidentHandler,
  handshake: handshakeHandler,
  pushToService: pushToServiceHandler,
};
