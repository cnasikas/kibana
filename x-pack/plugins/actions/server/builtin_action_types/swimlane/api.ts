/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import {
  CreateRecordApiHandlerArgs,
  ExternalServiceIncidentResponse,
  ExternalServiceApi,
  Incident,
  PushToServiceApiHandlerArgs,
  GetApplicationResponse,
  GetApplicationHandlerArgs,
} from './types';

const getApplicationHandler = async ({
  externalService,
}: GetApplicationHandlerArgs): Promise<GetApplicationResponse> => {
  return await externalService.getApplication();
};

const createRecordHandler = async ({
  externalService,
  params,
}: CreateRecordApiHandlerArgs): Promise<ExternalServiceIncidentResponse> => {
  return await externalService.createRecord({ incident: { ...params, externalId: null } });
};

const pushToServiceHandler = async ({
  externalService,
  params,
}: PushToServiceApiHandlerArgs): Promise<ExternalServiceIncidentResponse> => {
  const { comments } = params;
  let res: ExternalServiceIncidentResponse;
  const incident: Incident = params.incident;
  if (incident.externalId != null) {
    res = await externalService.updateRecord({
      incidentId: incident.externalId,
      incident,
    });
  } else {
    res = await externalService.createRecord({ incident });
  }

  const createdDate = new Date().toISOString();

  if (comments && Array.isArray(comments) && comments.length > 0) {
    for (const currentComment of comments) {
      await externalService.createComment({
        incidentId: res.id,
        comment: currentComment,
        createdDate,
      });
    }
  }

  return res;
};

export const api: ExternalServiceApi = {
  getApplication: getApplicationHandler,
  createRecord: createRecordHandler,
  pushToService: pushToServiceHandler,
};
