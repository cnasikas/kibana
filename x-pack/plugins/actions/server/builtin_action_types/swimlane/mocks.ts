/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { ExternalService } from './types';

export const applicationFields = [
  {
    id: 'adnjls',
    name: 'Alert Source',
    key: 'alert-source',
    fieldType: 'text',
  },
  {
    id: 'adnlas',
    name: 'Severity',
    key: 'severity',
    fieldType: 'text',
  },
  {
    id: 'adnfls',
    name: 'Rule Name',
    key: 'rule-name',
    fieldType: 'text',
  },
  {
    id: 'a6sst',
    name: 'Case Id',
    key: 'case-id-name',
    fieldType: 'text',
  },
  {
    id: 'a6fst',
    name: 'Case Name',
    key: 'case-name',
    fieldType: 'text',
  },
  {
    id: 'a6fdf',
    name: 'Comments',
    key: 'comments',
    fieldType: 'notes',
  },
  {
    id: 'a6fde',
    name: 'Description',
    key: 'description',
    fieldType: 'text',
  },
];

export const mappings = {
  alertSourceConfig: applicationFields[0],
  severityConfig: applicationFields[1],
  ruleNameConfig: applicationFields[2],
  caseIdConfig: applicationFields[3],
  caseNameConfig: applicationFields[4],
  commentsConfig: applicationFields[5],
  descriptionConfig: applicationFields[6],
};

export const getApplicationResponse = { fields: applicationFields };

export const recordResponseCreate = {
  id: '123456',
  title: 'neato',
  url: 'swimlane.com',
};

export const recordResponseUpdate = {
  id: '98765',
  title: 'not neato',
  url: 'laneswim.com',
};

export const commentResponse = {
  id: '123456',
};

const createMock = (): jest.Mocked<ExternalService> => {
  return {
    createComment: jest.fn().mockImplementation(() => Promise.resolve(commentResponse)),
    createRecord: jest.fn().mockImplementation(() => Promise.resolve(recordResponseCreate)),
    updateRecord: jest.fn().mockImplementation(() => Promise.resolve(recordResponseUpdate)),
  };
};

const externalServiceMock = {
  create: createMock,
};

const executorParams = {
  alertName: 'alert-name',
  alertSource: 'alert-source',
  caseId: 'case-id',
  caseName: 'case-name',
  comments: 'comments',
  severity: 'severity',
};

const apiParams = {
  ...executorParams,
};

export { externalServiceMock, executorParams, apiParams };
