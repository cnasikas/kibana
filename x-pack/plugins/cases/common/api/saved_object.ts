/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import * as rt from 'io-ts';

import { either } from 'fp-ts/lib/Either';

export const NumberFromString = new rt.Type<number, string, unknown>(
  'NumberFromString',
  rt.number.is,
  (u, c) =>
    either.chain(rt.string.validate(u, c), (s) => {
      const n = +s;
      return isNaN(n) ? rt.failure(u, c, 'cannot parse to a number') : rt.success(n);
    }),
  String
);

export const PaginationSchema = new rt.Type<
  { page?: number; perPage?: number },
  { page?: number; perPage?: number },
  unknown
>(
  'Test',
  rt.partial({ page: rt.number, perPage: rt.number }).is,
  (u, c) =>
    either.chain(rt.partial({ page: rt.number, perPage: rt.number }).validate(u, c), (params) => {
      if (params.page == null && params.perPage) {
        return rt.success(params);
      }

      const pageAsNumber = params.page ?? 0;
      const perPageAsNumber = params.perPage ?? 0;

      if (Math.max(pageAsNumber, pageAsNumber * perPageAsNumber) > 10) {
        return rt.failure(
          u,
          c,
          `The number of documents is too high. Paginating through more than ${10} documents is not possible.`
        );
      }

      return rt.success(params);
    }),
  rt.identity
);
