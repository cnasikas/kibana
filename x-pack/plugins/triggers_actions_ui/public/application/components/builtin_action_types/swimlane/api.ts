/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

export async function getApplication({
  signal,
  url,
  appId,
  apiToken,
}: {
  signal: AbortSignal;
  url: string;
  appId: string;
  apiToken: string;
}): Promise<Record<string, any>> {
  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
    'Private-Token': `${apiToken}`,
  };

  const urlWithoutTrailingSlash = url.endsWith('/') ? url.slice(0, -1) : url;
  const apiUrl = urlWithoutTrailingSlash.endsWith('api')
    ? urlWithoutTrailingSlash
    : urlWithoutTrailingSlash + '/api';
  const applicationUrl = `${apiUrl}/app/{appId}`;

  const getApplicationUrl = (id: string) => applicationUrl.replace('{appId}', id);
  try {
    const response = await window.fetch(getApplicationUrl(appId), {
      method: 'GET',
      headers,
      signal,
    });

    if (!response.ok) {
      throw new Error(
        `Received status: ${response.status} when attempting to get application with id: ${appId}`
      );
    }

    return await response.json();
  } catch (error) {
    throw new Error(`Unable to get application with id ${appId}. Error: ${error.message}`);
  }
}
