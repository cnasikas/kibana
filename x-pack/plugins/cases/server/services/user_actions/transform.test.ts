/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import {
  legacyTransformFindResponseToExternalModel,
  transformFindResponseToExternalModel,
  transformToExternalModel,
} from './transform';
import { createSOFindResponse } from '../test_utils';
import {
  createUserActionFindSO,
  createConnectorUserAction,
  createUserActionSO,
  updateConnectorUserAction,
  pushConnectorUserAction,
  createCaseUserAction,
  createPersistableStateUserAction,
  createExternalReferenceUserAction,
  testConnectorId,
} from './test_utils';
import { createPersistableStateAttachmentTypeRegistryMock } from '../../attachment_framework/mocks';
import type { SavedObjectsFindResponse } from '@kbn/core-saved-objects-api-server';
import type { ConnectorUserAction } from '../../../common/api';
import { Actions } from '../../../common/api';
import type { UserActionPersistedAttributes } from '../../common/types/user_actions';
import type { SavedObject } from '@kbn/core/server';

describe('transform', () => {
  const persistableStateAttachmentTypeRegistry = createPersistableStateAttachmentTypeRegistryMock();

  describe('action_id', () => {
    it('legacyTransformFindResponseToExternalModel sets action_id correctly to the saved object id', () => {
      const userAction = {
        ...createUserActionSO({ action: Actions.create, commentId: '5' }),
      };

      const transformed = legacyTransformFindResponseToExternalModel(
        createSOFindResponse([createUserActionFindSO(userAction)]),
        persistableStateAttachmentTypeRegistry
      );

      expect(transformed.saved_objects[0].attributes.action_id).toEqual('100');
    });

    it('transformFindResponseToExternalModel does not set action_id', () => {
      const userAction = {
        ...createUserActionSO({ action: Actions.create, commentId: '5' }),
      };

      const transformed = transformFindResponseToExternalModel(
        createSOFindResponse([createUserActionFindSO(userAction)]),
        persistableStateAttachmentTypeRegistry
      );

      expect(transformed.saved_objects[0].attributes).not.toHaveProperty('action_id');
    });
  });

  describe('case_id', () => {
    describe('legacyTransformFindResponseToExternalModel', () => {
      it('sets case_id correctly when it finds the reference', () => {
        const userAction = createConnectorUserAction();

        const transformed = legacyTransformFindResponseToExternalModel(
          createSOFindResponse([createUserActionFindSO(userAction)]),
          persistableStateAttachmentTypeRegistry
        );

        expect(transformed.saved_objects[0].attributes.case_id).toEqual('1');
      });

      it('sets case_id to an empty string when it cannot find the reference', () => {
        const userAction = {
          ...createConnectorUserAction(),
          references: [],
        };
        const transformed = legacyTransformFindResponseToExternalModel(
          createSOFindResponse([createUserActionFindSO(userAction)]),
          persistableStateAttachmentTypeRegistry
        );

        expect(transformed.saved_objects[0].attributes.case_id).toEqual('');
      });
    });

    describe('transformFindResponseToExternalModel', () => {
      it('does not set the case_id when the reference exists', () => {
        const userAction = createConnectorUserAction();

        const transformed = transformFindResponseToExternalModel(
          createSOFindResponse([createUserActionFindSO(userAction)]),
          persistableStateAttachmentTypeRegistry
        );

        expect(transformed.saved_objects[0].attributes).not.toHaveProperty('case_id');
      });

      it('does not set the case_id when the reference does not exist', () => {
        const userAction = {
          ...createConnectorUserAction(),
          references: [],
        };

        const transformed = transformFindResponseToExternalModel(
          createSOFindResponse([createUserActionFindSO(userAction)]),
          persistableStateAttachmentTypeRegistry
        );

        expect(transformed.saved_objects[0].attributes).not.toHaveProperty('case_id');
      });
    });
  });

  describe.each([
    [transformFindResponseToExternalModel.name, transformFindResponseToExternalModel],
    [legacyTransformFindResponseToExternalModel.name, legacyTransformFindResponseToExternalModel],
  ])('%s', (functionName, transformer) => {
    it('does not populate the ids when the response is an empty array', () => {
      expect(transformer(createSOFindResponse([]), persistableStateAttachmentTypeRegistry))
        .toMatchInlineSnapshot(`
        Object {
          "page": 1,
          "per_page": 0,
          "saved_objects": Array [],
          "total": 0,
        }
      `);
    });

    it('preserves the saved object fields and attributes when inject the ids', () => {
      const transformed = transformer(
        createSOFindResponse([createUserActionFindSO(createConnectorUserAction())]),
        persistableStateAttachmentTypeRegistry
      );

      expect(transformed).toMatchSnapshot();
    });

    it('populates the payload.connector.id for multiple user actions', () => {
      const transformed = transformer(
        createSOFindResponse([
          createUserActionFindSO(createConnectorUserAction()),
          createUserActionFindSO(createConnectorUserAction()),
        ]),
        persistableStateAttachmentTypeRegistry
      ) as SavedObjectsFindResponse<ConnectorUserAction>;

      expect(transformed.saved_objects[0].attributes.payload.connector.id).toEqual('1');
      expect(transformed.saved_objects[1].attributes.payload.connector.id).toEqual('1');
    });

    describe('reference ids', () => {
      it('sets comment_id to null when it cannot find the reference', () => {
        const userAction = {
          ...createUserActionSO({ action: Actions.create, commentId: '5' }),
          references: [],
        };
        const transformed = transformer(
          createSOFindResponse([createUserActionFindSO(userAction)]),
          persistableStateAttachmentTypeRegistry
        );

        expect(transformed.saved_objects[0].attributes.comment_id).toBeNull();
      });

      it('sets comment_id correctly when it finds the reference', () => {
        const userAction = createUserActionSO({
          action: Actions.create,
          commentId: '5',
        });

        const transformed = transformer(
          createSOFindResponse([createUserActionFindSO(userAction)]),
          persistableStateAttachmentTypeRegistry
        );

        expect(transformed.saved_objects[0].attributes.comment_id).toEqual('5');
      });
    });

    describe('persistable state attachments', () => {
      it('populates the persistable state', () => {
        const transformed = transformer(
          createSOFindResponse([createUserActionFindSO(createPersistableStateUserAction())]),
          persistableStateAttachmentTypeRegistry
        ) as SavedObjectsFindResponse<ConnectorUserAction>;

        expect(transformed).toMatchSnapshot();
      });
    });

    describe('external references', () => {
      it('populates the external references attributes', () => {
        const transformed = transformer(
          createSOFindResponse([createUserActionFindSO(createExternalReferenceUserAction())]),
          persistableStateAttachmentTypeRegistry
        ) as SavedObjectsFindResponse<ConnectorUserAction>;

        expect(transformed).toMatchSnapshot();
      });
    });
  });

  describe('connectors', () => {
    describe('create connector', () => {
      const userAction = createConnectorUserAction();
      testConnectorId(persistableStateAttachmentTypeRegistry, userAction, 'connector.id');
    });

    describe('update connector', () => {
      const userAction = updateConnectorUserAction();
      testConnectorId(persistableStateAttachmentTypeRegistry, userAction, 'connector.id');
    });

    describe('push connector', () => {
      const userAction = pushConnectorUserAction();
      testConnectorId(
        persistableStateAttachmentTypeRegistry,
        userAction,
        'externalService.connector_id',
        '100'
      );
    });

    describe('create case', () => {
      const userAction = createCaseUserAction();
      testConnectorId(persistableStateAttachmentTypeRegistry, userAction, 'connector.id');
    });
  });

  describe('transformToExternalModel', () => {
    it('returns the correct fields', () => {
      const transformed = transformToExternalModel(
        createConnectorUserAction(),
        persistableStateAttachmentTypeRegistry
      );

      expect(transformed.attributes).toMatchInlineSnapshot(`
        Object {
          "action": "create",
          "comment_id": null,
          "created_at": "abc",
          "created_by": Object {
            "email": "a",
            "full_name": "abc",
            "username": "b",
          },
          "owner": "securitySolution",
          "payload": Object {
            "connector": Object {
              "fields": Object {
                "issueType": "bug",
                "parent": "2",
                "priority": "high",
              },
              "id": "1",
              "name": ".jira",
              "type": ".jira",
            },
          },
          "type": "connector",
        }
      `);
    });

    it('preserves the saved object fields and attributes when inject the ids', () => {
      const transformed = transformToExternalModel(
        createConnectorUserAction(),
        persistableStateAttachmentTypeRegistry
      );

      expect(transformed).toMatchInlineSnapshot(`
        Object {
          "attributes": Object {
            "action": "create",
            "comment_id": null,
            "created_at": "abc",
            "created_by": Object {
              "email": "a",
              "full_name": "abc",
              "username": "b",
            },
            "owner": "securitySolution",
            "payload": Object {
              "connector": Object {
                "fields": Object {
                  "issueType": "bug",
                  "parent": "2",
                  "priority": "high",
                },
                "id": "1",
                "name": ".jira",
                "type": ".jira",
              },
            },
            "type": "connector",
          },
          "id": "100",
          "references": Array [
            Object {
              "id": "1",
              "name": "associated-cases",
              "type": "cases",
            },
            Object {
              "id": "1",
              "name": "connectorId",
              "type": "action",
            },
          ],
          "type": "cases-user-actions",
        }
      `);
    });

    it('sets comment_id to null when it cannot find the reference', () => {
      const userAction = {
        ...createUserActionSO({ action: Actions.create, commentId: '5' }),
        references: [],
      };
      const transformed = transformToExternalModel(
        userAction,
        persistableStateAttachmentTypeRegistry
      );

      expect(transformed.attributes.comment_id).toBeNull();
    });

    it('sets comment_id correctly when it finds the reference', () => {
      const userAction = createUserActionSO({
        action: Actions.create,
        commentId: '5',
      });

      const transformed = transformToExternalModel(
        userAction,
        persistableStateAttachmentTypeRegistry
      );

      expect(transformed.attributes.comment_id).toEqual('5');
    });

    const userActions = [
      ['create connector', createConnectorUserAction()],
      ['update connector', updateConnectorUserAction()],
      ['push connector', pushConnectorUserAction()],
      ['create case', createCaseUserAction()],
      ['persistable state', createPersistableStateUserAction()],
      ['external reference', createExternalReferenceUserAction()],
      [
        'description',
        createUserActionSO({
          action: Actions.create,
          payload: { description: 'test' },
          type: 'description',
        }),
      ],
    ] as Array<[string, SavedObject<UserActionPersistedAttributes>]>;

    it.each(userActions)('formats the payload correctly for: %s', (_, userAction) => {
      const transformed = transformToExternalModel(
        userAction,
        persistableStateAttachmentTypeRegistry
      );

      expect(transformed).toMatchSnapshot();
    });
  });
});
