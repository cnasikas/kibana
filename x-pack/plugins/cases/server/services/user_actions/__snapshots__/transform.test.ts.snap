// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform legacyTransformFindResponseToExternalModel external references populates the external references attributes 1`] = `
Object {
  "page": 1,
  "per_page": 1,
  "saved_objects": Array [
    Object {
      "attributes": Object {
        "action": "create",
        "action_id": "100",
        "case_id": "1",
        "comment_id": "external-reference-test-id",
        "created_at": "abc",
        "created_by": Object {
          "email": "a",
          "full_name": "abc",
          "username": "b",
        },
        "owner": "securitySolution",
        "payload": Object {
          "comment": Object {
            "externalReferenceAttachmentTypeId": ".test",
            "externalReferenceId": "my-id",
            "externalReferenceMetadata": null,
            "externalReferenceStorage": Object {
              "soType": "test-so",
              "type": "savedObject",
            },
            "owner": "securitySolution",
            "type": "externalReference",
          },
        },
        "type": "comment",
      },
      "id": "100",
      "references": Array [
        Object {
          "id": "my-id",
          "name": "externalReferenceId",
          "type": "test-so",
        },
        Object {
          "id": "1",
          "name": "associated-cases",
          "type": "cases",
        },
        Object {
          "id": "external-reference-test-id",
          "name": "associated-cases-comments",
          "type": "cases-comments",
        },
      ],
      "score": 0,
      "type": "cases-user-actions",
    },
  ],
  "total": 1,
}
`;

exports[`transform legacyTransformFindResponseToExternalModel persistable state attachments populates the persistable state 1`] = `
Object {
  "page": 1,
  "per_page": 1,
  "saved_objects": Array [
    Object {
      "attributes": Object {
        "action": "create",
        "action_id": "100",
        "case_id": "1",
        "comment_id": "persistable-state-test-id",
        "created_at": "abc",
        "created_by": Object {
          "email": "a",
          "full_name": "abc",
          "username": "b",
        },
        "owner": "securitySolution",
        "payload": Object {
          "comment": Object {
            "owner": "securitySolutionFixture",
            "persistableStateAttachmentState": Object {
              "foo": "foo",
              "injectedId": "testRef",
            },
            "persistableStateAttachmentTypeId": ".test",
            "type": "persistableState",
          },
        },
        "type": "comment",
      },
      "id": "100",
      "references": Array [
        Object {
          "id": "testRef",
          "name": "myTestReference",
          "type": "test-so",
        },
        Object {
          "id": "1",
          "name": "associated-cases",
          "type": "cases",
        },
        Object {
          "id": "persistable-state-test-id",
          "name": "associated-cases-comments",
          "type": "cases-comments",
        },
      ],
      "score": 0,
      "type": "cases-user-actions",
    },
  ],
  "total": 1,
}
`;

exports[`transform legacyTransformFindResponseToExternalModel preserves the saved object fields and attributes when inject the ids 1`] = `
Object {
  "page": 1,
  "per_page": 1,
  "saved_objects": Array [
    Object {
      "attributes": Object {
        "action": "create",
        "action_id": "100",
        "case_id": "1",
        "comment_id": null,
        "created_at": "abc",
        "created_by": Object {
          "email": "a",
          "full_name": "abc",
          "username": "b",
        },
        "owner": "securitySolution",
        "payload": Object {
          "connector": Object {
            "fields": Object {
              "issueType": "bug",
              "parent": "2",
              "priority": "high",
            },
            "id": "1",
            "name": ".jira",
            "type": ".jira",
          },
        },
        "type": "connector",
      },
      "id": "100",
      "references": Array [
        Object {
          "id": "1",
          "name": "associated-cases",
          "type": "cases",
        },
        Object {
          "id": "1",
          "name": "connectorId",
          "type": "action",
        },
      ],
      "score": 0,
      "type": "cases-user-actions",
    },
  ],
  "total": 1,
}
`;

exports[`transform transformFindResponseToExternalModel external references populates the external references attributes 1`] = `
Object {
  "page": 1,
  "per_page": 1,
  "saved_objects": Array [
    Object {
      "attributes": Object {
        "action": "create",
        "comment_id": "external-reference-test-id",
        "created_at": "abc",
        "created_by": Object {
          "email": "a",
          "full_name": "abc",
          "username": "b",
        },
        "owner": "securitySolution",
        "payload": Object {
          "comment": Object {
            "externalReferenceAttachmentTypeId": ".test",
            "externalReferenceId": "my-id",
            "externalReferenceMetadata": null,
            "externalReferenceStorage": Object {
              "soType": "test-so",
              "type": "savedObject",
            },
            "owner": "securitySolution",
            "type": "externalReference",
          },
        },
        "type": "comment",
      },
      "id": "100",
      "references": Array [
        Object {
          "id": "my-id",
          "name": "externalReferenceId",
          "type": "test-so",
        },
        Object {
          "id": "1",
          "name": "associated-cases",
          "type": "cases",
        },
        Object {
          "id": "external-reference-test-id",
          "name": "associated-cases-comments",
          "type": "cases-comments",
        },
      ],
      "score": 0,
      "type": "cases-user-actions",
    },
  ],
  "total": 1,
}
`;

exports[`transform transformFindResponseToExternalModel persistable state attachments populates the persistable state 1`] = `
Object {
  "page": 1,
  "per_page": 1,
  "saved_objects": Array [
    Object {
      "attributes": Object {
        "action": "create",
        "comment_id": "persistable-state-test-id",
        "created_at": "abc",
        "created_by": Object {
          "email": "a",
          "full_name": "abc",
          "username": "b",
        },
        "owner": "securitySolution",
        "payload": Object {
          "comment": Object {
            "owner": "securitySolutionFixture",
            "persistableStateAttachmentState": Object {
              "foo": "foo",
              "injectedId": "testRef",
            },
            "persistableStateAttachmentTypeId": ".test",
            "type": "persistableState",
          },
        },
        "type": "comment",
      },
      "id": "100",
      "references": Array [
        Object {
          "id": "testRef",
          "name": "myTestReference",
          "type": "test-so",
        },
        Object {
          "id": "1",
          "name": "associated-cases",
          "type": "cases",
        },
        Object {
          "id": "persistable-state-test-id",
          "name": "associated-cases-comments",
          "type": "cases-comments",
        },
      ],
      "score": 0,
      "type": "cases-user-actions",
    },
  ],
  "total": 1,
}
`;

exports[`transform transformFindResponseToExternalModel preserves the saved object fields and attributes when inject the ids 1`] = `
Object {
  "page": 1,
  "per_page": 1,
  "saved_objects": Array [
    Object {
      "attributes": Object {
        "action": "create",
        "comment_id": null,
        "created_at": "abc",
        "created_by": Object {
          "email": "a",
          "full_name": "abc",
          "username": "b",
        },
        "owner": "securitySolution",
        "payload": Object {
          "connector": Object {
            "fields": Object {
              "issueType": "bug",
              "parent": "2",
              "priority": "high",
            },
            "id": "1",
            "name": ".jira",
            "type": ".jira",
          },
        },
        "type": "connector",
      },
      "id": "100",
      "references": Array [
        Object {
          "id": "1",
          "name": "associated-cases",
          "type": "cases",
        },
        Object {
          "id": "1",
          "name": "connectorId",
          "type": "action",
        },
      ],
      "score": 0,
      "type": "cases-user-actions",
    },
  ],
  "total": 1,
}
`;

exports[`transform transformToExternalModel formats the payload correctly for: create case 1`] = `
Object {
  "attributes": Object {
    "action": "create",
    "comment_id": null,
    "created_at": "abc",
    "created_by": Object {
      "email": "a",
      "full_name": "abc",
      "username": "b",
    },
    "owner": "securitySolution",
    "payload": Object {
      "assignees": undefined,
      "connector": Object {
        "fields": Object {
          "issueType": "bug",
          "parent": "2",
          "priority": "high",
        },
        "id": "1",
        "name": ".jira",
        "type": ".jira",
      },
      "description": "a desc",
      "owner": "securitySolution",
      "settings": Object {
        "syncAlerts": false,
      },
      "severity": "low",
      "status": "open",
      "tags": Array [],
      "title": "a title",
    },
    "type": "create_case",
  },
  "id": "100",
  "references": Array [
    Object {
      "id": "1",
      "name": "associated-cases",
      "type": "cases",
    },
    Object {
      "id": "1",
      "name": "connectorId",
      "type": "action",
    },
  ],
  "type": "cases-user-actions",
}
`;

exports[`transform transformToExternalModel formats the payload correctly for: create connector 1`] = `
Object {
  "attributes": Object {
    "action": "create",
    "comment_id": null,
    "created_at": "abc",
    "created_by": Object {
      "email": "a",
      "full_name": "abc",
      "username": "b",
    },
    "owner": "securitySolution",
    "payload": Object {
      "connector": Object {
        "fields": Object {
          "issueType": "bug",
          "parent": "2",
          "priority": "high",
        },
        "id": "1",
        "name": ".jira",
        "type": ".jira",
      },
    },
    "type": "connector",
  },
  "id": "100",
  "references": Array [
    Object {
      "id": "1",
      "name": "associated-cases",
      "type": "cases",
    },
    Object {
      "id": "1",
      "name": "connectorId",
      "type": "action",
    },
  ],
  "type": "cases-user-actions",
}
`;

exports[`transform transformToExternalModel formats the payload correctly for: description 1`] = `
Object {
  "attributes": Object {
    "action": "create",
    "comment_id": null,
    "created_at": "abc",
    "created_by": Object {
      "email": "a",
      "full_name": "abc",
      "username": "b",
    },
    "owner": "securitySolution",
    "payload": Object {
      "description": "test",
    },
    "type": "description",
  },
  "id": "100",
  "references": Array [
    Object {
      "id": "1",
      "name": "associated-cases",
      "type": "cases",
    },
  ],
  "type": "cases-user-actions",
}
`;

exports[`transform transformToExternalModel formats the payload correctly for: external reference 1`] = `
Object {
  "attributes": Object {
    "action": "create",
    "comment_id": "external-reference-test-id",
    "created_at": "abc",
    "created_by": Object {
      "email": "a",
      "full_name": "abc",
      "username": "b",
    },
    "owner": "securitySolution",
    "payload": Object {
      "comment": Object {
        "externalReferenceAttachmentTypeId": ".test",
        "externalReferenceId": "my-id",
        "externalReferenceMetadata": null,
        "externalReferenceStorage": Object {
          "soType": "test-so",
          "type": "savedObject",
        },
        "owner": "securitySolution",
        "type": "externalReference",
      },
    },
    "type": "comment",
  },
  "id": "100",
  "references": Array [
    Object {
      "id": "my-id",
      "name": "externalReferenceId",
      "type": "test-so",
    },
    Object {
      "id": "1",
      "name": "associated-cases",
      "type": "cases",
    },
    Object {
      "id": "external-reference-test-id",
      "name": "associated-cases-comments",
      "type": "cases-comments",
    },
  ],
  "type": "cases-user-actions",
}
`;

exports[`transform transformToExternalModel formats the payload correctly for: persistable state 1`] = `
Object {
  "attributes": Object {
    "action": "create",
    "comment_id": "persistable-state-test-id",
    "created_at": "abc",
    "created_by": Object {
      "email": "a",
      "full_name": "abc",
      "username": "b",
    },
    "owner": "securitySolution",
    "payload": Object {
      "comment": Object {
        "owner": "securitySolutionFixture",
        "persistableStateAttachmentState": Object {
          "foo": "foo",
          "injectedId": "testRef",
        },
        "persistableStateAttachmentTypeId": ".test",
        "type": "persistableState",
      },
    },
    "type": "comment",
  },
  "id": "100",
  "references": Array [
    Object {
      "id": "testRef",
      "name": "myTestReference",
      "type": "test-so",
    },
    Object {
      "id": "1",
      "name": "associated-cases",
      "type": "cases",
    },
    Object {
      "id": "persistable-state-test-id",
      "name": "associated-cases-comments",
      "type": "cases-comments",
    },
  ],
  "type": "cases-user-actions",
}
`;

exports[`transform transformToExternalModel formats the payload correctly for: push connector 1`] = `
Object {
  "attributes": Object {
    "action": "push_to_service",
    "comment_id": null,
    "created_at": "abc",
    "created_by": Object {
      "email": "a",
      "full_name": "abc",
      "username": "b",
    },
    "owner": "securitySolution",
    "payload": Object {
      "externalService": Object {
        "connector_id": "100",
        "connector_name": ".jira",
        "external_id": "100",
        "external_title": "awesome",
        "external_url": "http://www.google.com",
        "pushed_at": "2019-11-25T21:54:48.952Z",
        "pushed_by": Object {
          "email": "testemail@elastic.co",
          "full_name": "elastic",
          "username": "elastic",
        },
      },
    },
    "type": "pushed",
  },
  "id": "100",
  "references": Array [
    Object {
      "id": "1",
      "name": "associated-cases",
      "type": "cases",
    },
    Object {
      "id": "100",
      "name": "pushConnectorId",
      "type": "action",
    },
  ],
  "type": "cases-user-actions",
}
`;

exports[`transform transformToExternalModel formats the payload correctly for: update connector 1`] = `
Object {
  "attributes": Object {
    "action": "update",
    "comment_id": null,
    "created_at": "abc",
    "created_by": Object {
      "email": "a",
      "full_name": "abc",
      "username": "b",
    },
    "owner": "securitySolution",
    "payload": Object {
      "connector": Object {
        "fields": Object {
          "issueType": "bug",
          "parent": "2",
          "priority": "high",
        },
        "id": "1",
        "name": ".jira",
        "type": ".jira",
      },
    },
    "type": "connector",
  },
  "id": "100",
  "references": Array [
    Object {
      "id": "1",
      "name": "associated-cases",
      "type": "cases",
    },
    Object {
      "id": "1",
      "name": "connectorId",
      "type": "action",
    },
  ],
  "type": "cases-user-actions",
}
`;
